### PR TITLE
Address ArrayVariable exodus restart issue

### LINKED ISSUE #17121: ArrayVariable won't restart properly
## Bug Description

When ArrayVariable (variable with component > 1) is used, moose automatically creates multiple variable names based on the name of the ArrayVariable and the component index. For example, if the input reads

```
  [./v]
    order = CONSTANT
    family = MONOMIAL
    components = 3
    initial_condition = '12 13 14'
  [../]
```

moose automatically creates `v_0`, `v_1`, and `v_2` variables.

When restarting, moose tries to read variable `v` from the restart file, which does not contains `v`.
See error message as follows:

```
map_find() error: key "v" not found in file ../src/systems/system.C on line 1405
Stack frames: 13
0: 0   libmesh_opt.0.dylib                 0x0000000106b0d865 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 1157
1: 1   libmesh_opt.0.dylib                 0x0000000106b0a55e libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 270
2: 2   libmesh_opt.0.dylib                 0x0000000107165b30 std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned short, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, unsigned short> > >::mapped_type const& libMesh::Utility::map_find<std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned short, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, unsigned short> > >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*)0>(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned short, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, unsigned short> > > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*, int) + 688
3: 3   libmesh_opt.0.dylib                 0x000000010716586c libMesh::System::variable_number(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const + 28
4: 4   libmesh_opt.0.dylib                 0x0000000106def3fb libMesh::ExodusII_IO::copy_elemental_solution(libMesh::System&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned int) + 43
5: 5   libmoose-opt.0.dylib                0x0000000105f2ca2d SystemBase::copyVars(libMesh::ExodusII_IO&) + 781
6: 6   libmoose-opt.0.dylib                0x0000000105b524a3 FEProblemBase::initialSetup() + 2531
7: 7   libmoose-opt.0.dylib                0x00000001058be604 Transient::init() + 1764
8: 8   libmoose-opt.0.dylib                0x000000010627198b MooseApp::executeExecutioner() + 75
9: 9   libmoose-opt.0.dylib                0x00000001062721b4 MooseApp::run() + 244
10: 10  sam-opt                             0x0000000104606b7c main + 140
11: 11  libdyld.dylib                       0x00007fff58a453d5 start + 1
12: 12  ???                                 0x0000000000000003 0x0 + 3
[0] ./include/libmesh/utility.h, line 154, compiled Dec 11 2020 at 10:49:02
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```

## Steps to Reproduce

Restart a simulation with ArrayVariable input as described above will trigger the error.

## Impact

Limit code restart capability if ArrayVariable is used.

### PR DESCRIPTION
**Pull request by lingzou**
Closes #17121

- In the `SystemBase::copyVars` function, add a logic to determine if the variable is an ArrayVariable by checking its size, i.e., `count()`.
- If it is an ArrayVariable, copy one by one, e.g., `Var_01`, `Var_02`...`Var_0N`
- There is however a defect here, the source var and dest var are assumed to have the same name.

### CODE REVIEW

File: framework/src/systems/SystemBase.C [hunk 1]
```diff
@@ -1222,11 +1222,27 @@ SystemBase::copyVars(ExodusII_IO & io)
...
-      if (getVariable(0, vci._dest_name).isNodal())
-        io.copy_nodal_solution(system(), vci._dest_name, vci._source_name, timestep);
-
+      if (getVariable(0, vci._dest_name).count() > 1)
+      {
+        ArrayMooseVariable & var = getFieldVariable<RealEigenVector>(0, vci._dest_name);
+        if (getVariable(0, vci._dest_name).isNodal())
+        {
+          for(unsigned int i = 0; i < getVariable(0, vci._dest_name).count(); i++)
+            io.copy_nodal_solution(system(), var.componentName(i), var.componentName(i), timestep);
```
Comment (YaqiWang) [inline framework/src/systems/SystemBase.C:1231]: You can call `SubProblem::arrayVariableComponent(vci._source_name, i)` here for the source variable names in exodus file.

**Review by YaqiWang — CHANGES_REQUESTED** [review]
Only have one comment which @lingzou has been aware of.

File: framework/src/systems/SystemBase.C (hunk 1 above)
Comment (lingzou) [reply framework/src/systems/SystemBase.C]: @YaqiWang I have used your suggestion in my latest commit.

**Comment by lingzou** [comment]
@loganharbour Do I remove WIP in the title, and let it go through regression test?

**Comment by milljm** [comment]
> @loganharbour Do I remove WIP in the title, and let it go through regression test?

Indeed, as the Civet log returns this:
```pre
2021-03-23 14:07:24:INFO:Ignoring work in progress PR: WIP: Address ArrayVariable restart issue [need suggestions to improve]
```

**Review by YaqiWang — DISMISSED** [review]
Looks good to me.

**Comment by YaqiWang** [comment]
@loganharbour this can be merged.

File: framework/src/systems/SystemBase.C [hunk 2]
```diff
@@ -1222,11 +1222,25 @@ SystemBase::copyVars(ExodusII_IO & io)
...
+        ArrayMooseVariable & var = getFieldVariable<RealEigenVector>(0, vci._dest_name);
+        for (unsigned int i = 0; i < getVariable(0, vci._dest_name).count(); i++)
+        {
+          std::string source_var = _subproblem.arrayVariableComponent(vci._source_name, i);
+          if (getVariable(0, vci._dest_name).isNodal())
+            io.copy_nodal_solution(system(), var.componentName(i), source_var, timestep);
+          else
+            io.copy_elemental_solution(system(), var.componentName(i), source_var, timestep);
+        }
+      }
```
Comment (loganharbour) [inline framework/src/systems/SystemBase.C:1236]:
```suggestion
      const auto & var = getVariable(0, vci._dest_name);
      if (var.count() > 1) // array variable
      {
        const auto & array_var = getFieldVariable<RealEigenVector>(0, vci._dest_name);
        for (MooseIndex(var.count()) i = 0; i < var.count(); ++i)
        {
          const auto exodus_var = _subproblem.arrayVariableComponent(vci._source_name, i);
          const auto system_var = array_var.componentName(i);
          if (var.isNodal())
            io.copy_nodal_solution(system(), exodus_var, system_var, timestep);
          else
            io.copy_elemental_solution(system(), exodus_var, system_var, timestep);
        }
      }
```

File: framework/src/systems/SystemBase.C [hunk 3]
```diff
@@ -1222,11 +1222,25 @@ SystemBase::copyVars(ExodusII_IO & io)
...
       else
-        io.copy_elemental_solution(system(), vci._dest_name, vci._source_name, timestep);
+      {
+        if (getVariable(0, vci._dest_name).isNodal())
+          io.copy_nodal_solution(system(), vci._dest_name, vci._source_name, timestep);
+        else
+          io.copy_elemental_solution(system(), vci._dest_name, vci._source_name, timestep);
```
Comment (loganharbour) [inline framework/src/systems/SystemBase.C:1242]:
```suggestion
        if (var.isNodal())
          io.copy_nodal_solution(system(), vci._dest_name, vci._source_name, timestep);
        else
          io.copy_elemental_solution(system(), vci._dest_name, vci._source_name, timestep);
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
Just minor style changes.

**Comment by loganharbour** [comment]
Can you add a test?

**Comment by lingzou** [comment]
> Can you add a test?

Sure!

**Comment by lingzou** [comment]
@loganharbour I need some help here to make the recover test working. I basically developed the two tests following the test case in restart/new_dt, and tried different things to make the recover work, but not successful. Could you please take a look at what went wrong here?

**Comment by loganharbour** [comment]
Sure @lingzou, I'm talking a look. The issue is that you can't run a restart with the `--recover` flag, so we can easily set the restart test to not run with recover by setting `recover = false` in the test spec. I'll put up a new commit with this change in addition to a few other cleanups.

There is one issue, though: this test is not actually calling the new lines of code that you added. Looking into that now.

**Review by loganharbour — APPROVED** [review]
Approval for @lingzou's commits.

**Comment by lingzou** [comment]
@loganharbour Thanks for the help!

**Review by zachmprince — APPROVED** [review]

**Pull request merged**